mindmeld/components/dialogue.py

```diff
@@ -1056,6 +1064,16 @@ async def invoke_async(self, request, responder):
         await self.invoke(request, responder)
 
 
+class DialogueResponderSchema(Schema):
+    frame = fields.Dict()
+    params = fields.Nested(ParamsSchema)
+    history = fields.List(fields.Dict())
```

Comment:
This is a good point. However, the history field is not entirely a DialogueResponder object since certain fields, namely responder object's history field and the responder object's request object's history field are removed to prevent circular references of the history data. This is why I left it as a dict.

Reply:
So I tried to implement the history as a list of responders since it has some nice guarantees on what is allowed in the history, but the main issue I faced is that the request object expects a history made of a tuple of immutable maps while the responder would have a history of responders. We could change the schema such that the request would deserialize to a history of responders, but that is a backwards incompatible change since the interface in the handlers to the request history has changed. To summarize, the incompatibility of the immutable map schema for history in request and the history being a list of responders in the responder causes serialization issues.

Since the request object needs to return an immutable maps for history, I unfortunately cannot guarantee type validation for the  history object :(

Reply:
Actually, I found a way of doing strict validation of the history object in the responder class by deserializing each element to a responder object and then serializing it again. This is not the best solution and reduced performance, but it guarantees us that the history elem is a serialized version of the responder object.